arrow: set default memset for when the platform doesn't set one

amitsadaphule opened the pull request:
Pulled the fix from apache/arrow commit # 89080e4.

On pcc64le platform, due to the missing default, the following
tests were failing with "panic:runtime error" due to SIGSEGV:
  pkg/sql/colexec:TestSupportedSQLTypesIntegration
  pkg/col/colserde:TestArrowBatchConverterRandom
  pkg/sql/logictest:TestLogic/fakedist/aggregate/other
Relevant part of the test failure log pasted below:
  panic: runtime error: invalid memory address or nil pointer dereference
  [signal SIGSEGV: segmentation violation code=0x. addr=0x. pc=0x........]

petermattis commented:
Hi @amitsadaphule. Unlike `github.com/cockroachdb/rocksdb`, we don't do PRs against the `vendored` repo. Instead, you run `bin/dep ensure -update` from the `cockroachdb/cockroach` repo. We're near enough to the release of the 20.1 cycle that it would be good to only update the arrow dependency. I believe this would be done by `bin/dep ensure -update github.com/apache/arrow`. See the instructions in https://github.com/cockroachdb/cockroach/blob/master/build/README.md for some more details about updating vendored dependencies.

petermattis commented:
@amitsadaphule Oops. I was mistaken. You will need this PR because you can't push branches to this repo. I think this PR contains too many changes, though. As I mentioned in my previous message, we should only be updating the minimum necessary at this point in the release cycle.

petermattis commented:
Hmm, the `vendor` pointer in the CockroachDB repo does not actually point to `master` in this repo. Opening a PR against `master` is going to be problematic.

dt commented:
@amitsadaphule I think the parent of your commit might have been too old -- you want to start with `vendor` checked out at the currently SHA used by cockroachdb/cockroach, i.e. 002b8d2f136035aef972ef67cf6d546840a0e5fd right now, and then make your change. It doesn't really matter what you PR against since the important thing is that we have a hosted ref that `cockroachdb/cockroach`'s submodule pointer can then use, so against master is OK (as long as we keep the parent correct).

All that said, note that we *cannot* make manual changes to files in `vendored` -- they will be overwritten by the next person to run `dep ensure`. If you want to pick up a fix in some library in here, we either need to bump that library version to a published ref upstream using `dep`, or, if we need some custom SHA, we'd need to switch to consume a fork of upstream and then bump to that, again using `dep`.

amitsadaphule commented:
@petermattis thanks for the feedback! Should I make a PR with reference to vendored master with only the arrow related commit then? If that's the case, I'm not sure whether cockroach builds properly with vendored master though. I added my changes on top of the latest snapshot of vendored submodule being used by cockroach master.

amitsadaphule commented:
@dt thanks for the feedback! Actually, the parent of my commit is correct, which is vendor checked out at the current SHA used by cockroachdb/cockroach (i.e. 002b8d2 right now).

About updating vendor using dep, I tried using `bin/dep ensure -update github.com/apache/arrow` as suggested by @petermattis . But that did not bring in the expected fix. The latest release/master of apache/arrow contains the fix though. Could you please let me know how to get the latest release of arrow using dep? I'll need to close this PR and raise a new one in that care, correct?

dt commented:
Hey @amitsadaphule thanks for tracking this down and for the PR!

The way we'd usually pick up an upstream fix would be `bin/dep ensure -update github.com/apache/arrow` which, right now, for me, picks up a pretty large set of upstream changes: https://github.com/apache/arrow/compare/338c62a2...f7d39232. If we want a _speciifc_ version of upstream, we can also add a constraint to `Gopkg.toml` like this before running that `ensure` to control what it picks: 
```
[[constraint]]
  name = "github.com/apache/arrow"
  version = "apache-arrow-0.16.0"
```

However As Peter mentioned above, given our proximity to release right now, we don't really want to pick up a big batch of upstream changes from Arrow -- there are hundreds of commits between what we're currently using and have been testing with and that patch. To pick up just that single patch, we would need to host a fork of upstream arrow where we could cherrypick just that patch on top of the SHA we're currently using.

Would you be okay with waiting for the 20.1 release to be cut and then we could just bump arrow to a current version on master? That would mean you'd need to use master builds of cockroach for ppc support until at least 20.2, but we're really trying to minimize changes right now ahead of 20.1 the imminent release.

amitsadaphule commented:
@dt thanks for the feedback! Yes, I understand your concern. I'm fine with waiting until 20.1 is released, since I'm anyway using v19.2.2 with pre-configured patches in the build process for ppc64le.

amitsadaphule commented:
@dt @petermattis I noticed that cockroach has moved past 20.1 (v20.1.1 released already). Would this be a good time to start exploring the options that you had suggested above?

dt commented:
Hey @amitsadaphule, thanks for following up on this!

Did the change that @asubiotto merged in https://github.com/cockroachdb/cockroach/pull/49184 address this?

amitsadaphule commented:
@dt thanks for the feedback! I built cockroach master ([this](https://github.com/cockroachdb/cockroach/commit/a8e821d13b2b623c38ba82c577536ecde47bd685) commit to be precise), confirmed that this change is part of the `vendor/github.com/apache/arrow` dependency now and executed the 3 tests which were failing earlier in the absence of the fix as mentioned below:
```
1. github.com/cockroachdb/cockroach/pkg/col/colserde

--- FAIL: TestArrowBatchConverterRandom (0.00s)
panic: runtime error: invalid memory address or nil pointer dereference [recovered]
panic: runtime error: invalid memory address or nil pointer dereference [recovered]
panic: runtime error: invalid memory address or nil pointer dereference
[signal SIGSEGV: segmentation violation code=0x1 addr=0x0 pc=0x10169178]

2. github.com/cockroachdb/cockroach/pkg/sql/colexec

--- FAIL: TestSupportedSQLTypesIntegration (0.00s)
panic: runtime error: invalid memory address or nil pointer dereference [recovered]
panic: runtime error: invalid memory address or nil pointer dereference [recovered]
panic: runtime error: invalid memory address or nil pointer dereference [recovered]
panic: runtime error: invalid memory address or nil pointer dereference
[signal SIGSEGV: segmentation violation code=0x1 addr=0x0 pc=0x10e4e448]

3. github.com/cockroachdb/cockroach/pkg/sql/logictest

=== RUN   TestLogic/fakedist/aggregate/other
rewrote:
CREATE TABLE kv (k INT8 PRIMARY KEY, v INT8, w INT8, s STRING, i INTERVAL, FAMILY (s), FAMILY (w, i), FAMILY (v));


rewrote:
CREATE TABLE a (n INT8 PRIMARY KEY, a INT8, b INT8, c STRING, INDEX a_idx (a), INDEX b_idx (b), FAMILY (c), FAMILY (b, a));
INSERT INTO a SELECT a, a, a % 3, 'foo' FROM ROWS FROM (generate_series(1, 10)) AS g (a);
SET enable_zigzag_join = true;


rewrote:
CREATE TABLE f (i INT8, v FLOAT8, FAMILY (i, v));


panic: runtime error: invalid memory address or nil pointer dereference
[signal SIGSEGV: segmentation violation code=0x1 addr=0x0 pc=0x11fecd48]
```
I confirmed that the issue is fixed now (logs below). So, closing this PR.

```
[root@bc1a3288a2b3 cockroach]# make test PKG=./pkg/sql/colexec TESTS=TestSupportedSQLTypesIntegration TESTFLAGS='-v -count=1'
Running make with -j8
GOPATH set to /root/go
mkdir -p lib
ln -sf /root/go/native/ppc64le-redhat-linux/geos/lib/lib{geos,geos_c}.so lib
GOFLAGS= go test   -tags ' make ppc64le_redhat_linux' -ldflags '-X github.com/cockroachdb/cockroach/pkg/build.typ=development -extldflags "" -X "github.com/cockroachdb/cockroach/pkg/build.tag=v20.2.0-alpha.00000000-2388-ga8e821d" -X "github.com/cockroachdb/cockroach/pkg/build.rev=a8e821d13b2b623c38ba82c577536ecde47bd685" -X "github.com/cockroachdb/cockroach/pkg/build.cgoTargetTriple=ppc64le-redhat-linux"  ' -run "TestSupportedSQLTypesIntegration"  -timeout 30m ./pkg/sql/colexec -v -count=1
I200617 09:36:58.355944 1 rand.go:85  Random seed: 8485394295487407090
coldata.BatchSize() is set to 3538
testing: warning: no tests to run
PASS
ok      github.com/cockroachdb/cockroach/pkg/sql/colexec        0.039s [no tests to run]


[root@bc1a3288a2b3 cockroach]# make test PKG=./pkg/col/colserde TESTS=TestArrowBatchConverterRandom TESTFLAGS='-v -count=1'
Running make with -j8
GOPATH set to /root/go
mkdir -p lib
ln -sf /root/go/native/ppc64le-redhat-linux/geos/lib/lib{geos,geos_c}.so lib
GOFLAGS= go test   -tags ' make ppc64le_redhat_linux' -ldflags '-X github.com/cockroachdb/cockroach/pkg/build.typ=development -extldflags "" -X "github.com/cockroachdb/cockroach/pkg/build.tag=v20.2.0-alpha.00000000-2388-ga8e821d" -X "github.com/cockroachdb/cockroach/pkg/build.rev=a8e821d13b2b623c38ba82c577536ecde47bd685" -X "github.com/cockroachdb/cockroach/pkg/build.cgoTargetTriple=ppc64le-redhat-linux"  ' -run "TestArrowBatchConverterRandom"  -timeout 30m ./pkg/col/colserde -v -count=1
I200617 09:37:38.840672 1 rand.go:85  Random seed: -5595349863785047535
=== RUN   TestArrowBatchConverterRandom
--- PASS: TestArrowBatchConverterRandom (0.00s)
PASS
ok      github.com/cockroachdb/cockroach/pkg/col/colserde       0.035s


[root@bc1a3288a2b3 cockroach]# make test PKG=./pkg/sql/logictest TESTS=TestLogic/fakedist/aggregate/other TESTFLAGS='-v -count=1'
Running make with -j8
GOPATH set to /root/go
mkdir -p lib
ln -sf /root/go/native/ppc64le-redhat-linux/geos/lib/lib{geos,geos_c}.so lib
GOFLAGS= go test   -tags ' make ppc64le_redhat_linux' -ldflags '-X github.com/cockroachdb/cockroach/pkg/build.typ=development -extldflags "" -X "github.com/cockroachdb/cockroach/pkg/build.tag=v20.2.0-alpha.00000000-2388-ga8e821d" -X "github.com/cockroachdb/cockroach/pkg/build.rev=a8e821d13b2b623c38ba82c577536ecde47bd685" -X "github.com/cockroachdb/cockroach/pkg/build.cgoTargetTriple=ppc64le-redhat-linux"  ' -run "TestLogic/fakedist/aggregate/other"  -timeout 30m ./pkg/sql/logictest -v -count=1
I200617 09:39:16.528394 1 rand.go:85  Random seed: 6565374648228686560
=== RUN   TestLogic
=== RUN   TestLogic/fakedist
=== RUN   TestLogic/fakedist/aggregate
=== PAUSE TestLogic/fakedist/aggregate
=== CONT  TestLogic/fakedist/aggregate
=== RUN   TestLogic/fakedist/aggregate/other
rewrote:
CREATE TABLE kv (k INT8 PRIMARY KEY, v INT8, w INT8, s STRING, i INTERVAL, FAMILY (k), FAMILY (w), FAMILY (i), FAMILY (s), FAMILY (v));


rewrote:
CREATE TABLE abc (a VARCHAR PRIMARY KEY, b FLOAT8, c BOOL, d DECIMAL, FAMILY (a), FAMILY (b), FAMILY (c), FAMILY (d));


rewrote:
CREATE TABLE mnop (m INT8 PRIMARY KEY, n FLOAT8, o DECIMAL, p INT8, FAMILY (m, p), FAMILY (n, o));


--- progress: testdata/logic_test/aggregate: 150 statements/queries
rewrote:
CREATE TABLE filter_test (k INT8, v INT8, mark BOOL, FAMILY (v), FAMILY (k, mark));


rewrote:
CREATE TABLE opt_test (k INT8 PRIMARY KEY, v INT8, INDEX v (v), FAMILY (k), FAMILY (v));


rewrote:
CREATE TABLE xor_bytes (a BYTES, b INT8, c INT8, FAMILY (c), FAMILY (a, b));


rewrote:
INSERT INTO ab(a, b) VALUES (1, 2), (3, 4);
CREATE TABLE xy (x STRING, y STRING, FAMILY (y), FAMILY (x));
INSERT INTO xy(x, y) VALUES ('a', 'b'), ('c', 'd');


rewrote:
CREATE TABLE group_ord (x INT8 PRIMARY KEY, y INT8, z INT8, INDEX foo (z), FAMILY (y, z, x));


--- done: testdata/logic_test/aggregate with config fakedist: 214 tests, 0 failures
--- total progress: 214 statements/queries
=== RUN   TestLogic/fakedist-vec-off
=== RUN   TestLogic/fakedist-vec-off/aggregate
=== PAUSE TestLogic/fakedist-vec-off/aggregate
=== CONT  TestLogic/fakedist-vec-off/aggregate
=== RUN   TestLogic/fakedist-vec-off/aggregate/other
rewrote:
CREATE TABLE kv (k INT8 PRIMARY KEY, v INT8, w INT8, s STRING, i INTERVAL, FAMILY (k), FAMILY (i, w, v), FAMILY (s));


rewrote:
CREATE TABLE abc (a VARCHAR PRIMARY KEY, b FLOAT8, c BOOL, d DECIMAL, FAMILY (a), FAMILY (c, d, b));


rewrote:
CREATE TABLE mnop (m INT8 PRIMARY KEY, n FLOAT8, o DECIMAL, p INT8, FAMILY (n), FAMILY (m, o, p));


--- progress: testdata/logic_test/aggregate: 151 statements/queries
rewrote:
CREATE TABLE filter_test (k INT8, v INT8, mark BOOL, FAMILY (v, k), FAMILY (mark));


rewrote:
CREATE TABLE opt_test (k INT8 PRIMARY KEY, v INT8, INDEX v (v), FAMILY (v), FAMILY (k));


rewrote:
CREATE TABLE xor_bytes (a BYTES, b INT8, c INT8, FAMILY (b), FAMILY (a), FAMILY (c));


rewrote:
INSERT INTO ab(a, b) VALUES (1, 2), (3, 4);
CREATE TABLE xy (x STRING, y STRING, FAMILY (x), FAMILY (y));
INSERT INTO xy(x, y) VALUES ('a', 'b'), ('c', 'd');


rewrote:
CREATE TABLE group_ord (x INT8 PRIMARY KEY, y INT8, z INT8, INDEX foo (z), FAMILY (x), FAMILY (z), FAMILY (y));


--- done: testdata/logic_test/aggregate with config fakedist-vec-off: 214 tests, 0 failures
--- total progress: 428 statements/queries
=== RUN   TestLogic/fakedist-vec-auto
=== RUN   TestLogic/fakedist-vec-auto/aggregate
=== PAUSE TestLogic/fakedist-vec-auto/aggregate
=== CONT  TestLogic/fakedist-vec-auto/aggregate
=== RUN   TestLogic/fakedist-vec-auto/aggregate/other
rewrote:
CREATE TABLE kv (k INT8 PRIMARY KEY, v INT8, w INT8, s STRING, i INTERVAL, FAMILY (s, v), FAMILY (i), FAMILY (w, k));


rewrote:
CREATE TABLE abc (a VARCHAR PRIMARY KEY, b FLOAT8, c BOOL, d DECIMAL, FAMILY (c, b), FAMILY (d), FAMILY (a));


rewrote:
CREATE TABLE mnop (m INT8 PRIMARY KEY, n FLOAT8, o DECIMAL, p INT8, FAMILY (m), FAMILY (o), FAMILY (p, n));


--- progress: testdata/logic_test/aggregate: 150 statements/queries
rewrote:
CREATE TABLE filter_test (k INT8, v INT8, mark BOOL, FAMILY (mark, k, v));


rewrote:
CREATE TABLE opt_test (k INT8 PRIMARY KEY, v INT8, INDEX v (v), FAMILY (k, v));


rewrote:
CREATE TABLE xor_bytes (a BYTES, b INT8, c INT8, FAMILY (c), FAMILY (b), FAMILY (a));


rewrote:
INSERT INTO ab(a, b) VALUES (1, 2), (3, 4);
CREATE TABLE xy (x STRING, y STRING, FAMILY (y, x));
INSERT INTO xy(x, y) VALUES ('a', 'b'), ('c', 'd');


rewrote:
CREATE TABLE group_ord (x INT8 PRIMARY KEY, y INT8, z INT8, INDEX foo (z), FAMILY (x), FAMILY (y, z));


--- done: testdata/logic_test/aggregate with config fakedist-vec-auto: 214 tests, 0 failures
--- total progress: 642 statements/queries
=== RUN   TestLogic/fakedist-vec-auto-disk
=== RUN   TestLogic/fakedist-vec-auto-disk/aggregate
=== PAUSE TestLogic/fakedist-vec-auto-disk/aggregate
=== CONT  TestLogic/fakedist-vec-auto-disk/aggregate
=== RUN   TestLogic/fakedist-vec-auto-disk/aggregate/other
rewrote:
CREATE TABLE kv (k INT8 PRIMARY KEY, v INT8, w INT8, s STRING, i INTERVAL, FAMILY (w), FAMILY (v), FAMILY (i, s), FAMILY (k));


rewrote:
CREATE TABLE abc (a VARCHAR PRIMARY KEY, b FLOAT8, c BOOL, d DECIMAL, FAMILY (a, c, d), FAMILY (b));


rewrote:
CREATE TABLE mnop (m INT8 PRIMARY KEY, n FLOAT8, o DECIMAL, p INT8, FAMILY (n), FAMILY (p, o, m));


--- progress: testdata/logic_test/aggregate: 151 statements/queries
rewrote:
CREATE TABLE filter_test (k INT8, v INT8, mark BOOL, FAMILY (mark, k, v));


rewrote:
CREATE TABLE opt_test (k INT8 PRIMARY KEY, v INT8, INDEX v (v), FAMILY (k, v));


rewrote:
CREATE TABLE xor_bytes (a BYTES, b INT8, c INT8, FAMILY (c), FAMILY (a), FAMILY (b));


rewrote:
INSERT INTO ab(a, b) VALUES (1, 2), (3, 4);
CREATE TABLE xy (x STRING, y STRING, FAMILY (x, y));
INSERT INTO xy(x, y) VALUES ('a', 'b'), ('c', 'd');


rewrote:
CREATE TABLE group_ord (x INT8 PRIMARY KEY, y INT8, z INT8, INDEX foo (z), FAMILY (y), FAMILY (z, x));


--- done: testdata/logic_test/aggregate with config fakedist-vec-auto-disk: 214 tests, 0 failures
--- total progress: 856 statements/queries
=== RUN   TestLogic/fakedist-metadata
=== RUN   TestLogic/fakedist-metadata/aggregate
=== PAUSE TestLogic/fakedist-metadata/aggregate
=== CONT  TestLogic/fakedist-metadata/aggregate
=== RUN   TestLogic/fakedist-metadata/aggregate/other
rewrote:
CREATE TABLE kv (k INT8 PRIMARY KEY, v INT8, w INT8, s STRING, i INTERVAL, FAMILY (w), FAMILY (i), FAMILY (v), FAMILY (k, s));


rewrote:
CREATE TABLE abc (a VARCHAR PRIMARY KEY, b FLOAT8, c BOOL, d DECIMAL, FAMILY (a, d, b, c));


rewrote:
CREATE TABLE mnop (m INT8 PRIMARY KEY, n FLOAT8, o DECIMAL, p INT8, FAMILY (m), FAMILY (n, p, o));


--- progress: testdata/logic_test/aggregate: 150 statements/queries
rewrote:
CREATE TABLE filter_test (k INT8, v INT8, mark BOOL, FAMILY (mark), FAMILY (k, v));


rewrote:
CREATE TABLE opt_test (k INT8 PRIMARY KEY, v INT8, INDEX v (v), FAMILY (v), FAMILY (k));


rewrote:
CREATE TABLE xor_bytes (a BYTES, b INT8, c INT8, FAMILY (c), FAMILY (a), FAMILY (b));


rewrote:
INSERT INTO ab(a, b) VALUES (1, 2), (3, 4);
CREATE TABLE xy (x STRING, y STRING, FAMILY (y), FAMILY (x));
INSERT INTO xy(x, y) VALUES ('a', 'b'), ('c', 'd');


rewrote:
CREATE TABLE group_ord (x INT8 PRIMARY KEY, y INT8, z INT8, INDEX foo (z), FAMILY (y, x, z));


--- done: testdata/logic_test/aggregate with config fakedist-metadata: 214 tests, 0 failures
--- total progress: 1070 statements/queries
=== RUN   TestLogic/fakedist-disk
=== RUN   TestLogic/fakedist-disk/aggregate
=== PAUSE TestLogic/fakedist-disk/aggregate
=== CONT  TestLogic/fakedist-disk/aggregate
=== RUN   TestLogic/fakedist-disk/aggregate/other
rewrote:
CREATE TABLE kv (k INT8 PRIMARY KEY, v INT8, w INT8, s STRING, i INTERVAL, FAMILY (k, i, v), FAMILY (w), FAMILY (s));


rewrote:
CREATE TABLE abc (a VARCHAR PRIMARY KEY, b FLOAT8, c BOOL, d DECIMAL, FAMILY (c, d), FAMILY (a, b));


rewrote:
CREATE TABLE mnop (m INT8 PRIMARY KEY, n FLOAT8, o DECIMAL, p INT8, FAMILY (n, m, o, p));


--- progress: testdata/logic_test/aggregate: 151 statements/queries
rewrote:
CREATE TABLE filter_test (k INT8, v INT8, mark BOOL, FAMILY (v), FAMILY (k), FAMILY (mark));


rewrote:
CREATE TABLE opt_test (k INT8 PRIMARY KEY, v INT8, INDEX v (v), FAMILY (k, v));


rewrote:
CREATE TABLE xor_bytes (a BYTES, b INT8, c INT8, FAMILY (b, c, a));


rewrote:
INSERT INTO ab(a, b) VALUES (1, 2), (3, 4);
CREATE TABLE xy (x STRING, y STRING, FAMILY (y), FAMILY (x));
INSERT INTO xy(x, y) VALUES ('a', 'b'), ('c', 'd');


rewrote:
CREATE TABLE group_ord (x INT8 PRIMARY KEY, y INT8, z INT8, INDEX foo (z), FAMILY (y), FAMILY (z, x));


--- done: testdata/logic_test/aggregate with config fakedist-disk: 214 tests, 0 failures
--- total progress: 1284 statements/queries
--- total: 1284 tests, 0 failures
--- PASS: TestLogic (29.40s)
    test_log_scope.go:77: test logs captured to: /tmp/logTestLogic961806487
    test_log_scope.go:58: use -show-logs to present logs inline
    --- PASS: TestLogic/fakedist (0.00s)
        --- PASS: TestLogic/fakedist/aggregate (6.09s)
            --- PASS: TestLogic/fakedist/aggregate/other (5.04s)
    --- PASS: TestLogic/fakedist-vec-off (0.00s)
        --- PASS: TestLogic/fakedist-vec-off/aggregate (4.04s)
            --- PASS: TestLogic/fakedist-vec-off/aggregate/other (3.20s)
    --- PASS: TestLogic/fakedist-vec-auto (0.00s)
        --- PASS: TestLogic/fakedist-vec-auto/aggregate (6.04s)
            --- PASS: TestLogic/fakedist-vec-auto/aggregate/other (4.95s)
    --- PASS: TestLogic/fakedist-vec-auto-disk (0.00s)
        --- PASS: TestLogic/fakedist-vec-auto-disk/aggregate (4.03s)
            --- PASS: TestLogic/fakedist-vec-auto-disk/aggregate/other (3.28s)
    --- PASS: TestLogic/fakedist-metadata (0.00s)
        --- PASS: TestLogic/fakedist-metadata/aggregate (5.03s)
            --- PASS: TestLogic/fakedist-metadata/aggregate/other (4.21s)
    --- PASS: TestLogic/fakedist-disk (0.00s)
        --- PASS: TestLogic/fakedist-disk/aggregate (4.03s)
            --- PASS: TestLogic/fakedist-disk/aggregate/other (3.18s)
PASS
ok      github.com/cockroachdb/cockroach/pkg/sql/logictest      29.450s

```
